### ISSUE TITLE
Minor issues creating SOAP channel in web UI

### ISSUE THREAD

**Issue by candlerb** [zatosource/zato#377]
(Maybe related to #376, but this is when creating SOAP channels via the Web UI)

* if I try to create a SOAP channel in the web UI but leave the name blank, I get a nasty backtrace instead of a validation error in the UI

~~~
Could not create the object, e:[Traceback (most recent call last): File "/opt/zato/2.0.1/zato-web-admin/src/zato/admin/web/views/http_soap.py", line 189, in create response = req.zato.client.invoke('zato.http-soap.create', _get_edit_create_message(req.POST)) File "/opt/zato/2.0.1/zato-web-admin/src/zato/admin/middleware.py", line 70, in invoke raise Exception(response.inner.text) Exception: {"zato_env": {"details": "Traceback (most recent call last):\n File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/connection\/http_soap\/channel.py\", line 198, in dispatch\n payload, worker_store, self.simple_io_config, post_data)\n File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/connection\/http_soap\/channel.py\", line 351, in handle\n params_priority=channel_item.params_pri)\n File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/service\/__init__.py\", line 426, in update_handle\n raise e\nZatoException: Expected elem:[name] not found in item:[{u'id': 536, u'name': u''}]\n", "result": "ZATO_ERROR", "cid": "K06FP8YV4N72DZK407602NCNGJDM"}} ]
~~~

* If I do specify all the parameters, then instead of getting back an 'OK' message then I get some JSON rendered directly in the browser. This is using Chrome 40 under OSX 10.9.5

~~~
{"message": "Successfully created the SOAP channel [ordering api1], check server logs for details", "id": 537, "transport": "soap"}
~~~

**Comment by dsuch** [comment]
Hi @candlerb,

this is actually what I consider a regression in Chrome but it would take ages to trace it back to a specific release and report it back upstream.

The workaround right now is to use FF.

**Comment by candlerb** [comment]
More serious problem: after creating the channel, the web interface barfs when I try to list channels <http://x.x.x.x:8183/zato/http-soap/?cluster=1&connection=channel&transport=soap>

~~~
{"zato_env": {"details": "Traceback (most recent call last):\n
  File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/connection\/http_soap\/channel.py\", line 198, in dispatch\n    payload, worker_store, self.simple_io_config, post_data)\n
  File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/connection\/http_soap\/channel.py\", line 351, in handle\n    params_priority=channel_item.params_pri)\n
  File \"\/opt\/zato\/2.0.1\/zato-server\/src\/zato\/server\/service\/__init__.py\", line 426, in update_handle\n
    raise e\n
ZatoException: Expected elem:[name] not found in item:[(['id', 'name', 'is_active', 'is_internal', 'transport', 'host', 'url_path', 'method', 'soap_action', 'soap_version', 'data_format', 'security_id', 'has_rbac', 'connection', u'ping_method', u'pool_size', u'merge_url_params_req', u'url_params_pri', u'params_pri', u'serialization_type', 'audit_enabled', 'audit_back_log', 'audit_max_payload', 'audit_repl_patt_type', 'timeout', 'sec_tls_ca_cert_id', u'sec_tls_ca_cert_name', 'sec_type', u'service_name', u'service_id', u'service_impl_name', u'security_name', u'username', u'password', u'password_type'], (536, u'', True, False, u'soap', None, u'\/api1', u'POST', u'sample.hello', u'1.1', u'xml', None, False, u'channel', u'HEAD', 20, True, u'qs-over-path', u'channel-params-over-msg', u'string', False, 1440, 0, u'json-pointer', 10, None, None, None, u'sample.hello', 378, u'sample.Hello', None, None, None, None))]\n
", "result": "ZATO_ERROR", "cid": "K06YPKDE3KQY92343PV1A434X2TK"}}
~~~

**Comment by dsuch** [comment]
There's a file called logs/admin.log (in both servers) - can you please post its contents around the time you issue the list request?

**Comment by candlerb** [comment]
It looks like my nameless service *did* somehow make it into the database:

~~~
sqlite> select * from http_soap where url_path='/api1';
536||1|0|channel|soap||/api1|POST|sample.hello|1.1|xml|HEAD|20|1|qs-over-path|channel-params-over-msg|0|1440|0|json-pointer|string|10||0|378|1|
537|ordering api1|1|0|channel|soap||/api1|POST|sample.hello|1.1|xml|HEAD|20|1|qs-over-path|channel-params-over-msg|0|1440|0|json-pointer|string|10||0|378|1|
sqlite> select * from http_soap where name is null;
sqlite> select * from http_soap where name='';
536||1|0|channel|soap||/api1|POST|sample.hello|1.1|xml|HEAD|20|1|qs-over-path|channel-params-over-msg|0|1440|0|json-pointer|string|10||0|378|1|
~~~

So I have deleted that row by hand and it's now OK. Would be better not to get into that state obviously.

**Comment by dsuch** [comment]
> Would be better not to get into that state obviously.

Right - this is related to the fact that Chrome apparently submitted an empty string because the frontend's JS validation didn't work in that browser. Had it been a None received on server side the request would have been rejected.

Again, support for Chrome must be brought to order.

**Comment by candlerb** [comment]
(If the browser is calling the public API then it seems to me that this is a failure of validation in the backend, since clients cannot be trusted)

Logs:

~~~
$ grep "^2015-02-25 22:13:32" server2/logs/admin.log
2015-02-25 22:13:32,158 - INFO - 8016:Dummy-1190 - zato_admin:22 - cid:[K062JEXADBJKNAA58DYYSKGQTS69], name:[zato.service.invoke], SIO request:[{u'name': u'zato.security.tls.ca-cert.get-list', u'data_format': u'json', u'id': u'', u'channel': u'invoke', u'async': False, u'payload': u'eyJjbHVzdGVyX2lkIjogMX0=\n', u'transport': None, u'expiration': 15}]
2015-02-25 22:13:32,160 - INFO - 8016:Dummy-1190 - zato_admin:22 - cid:[K062JEXADBJKNAA58DYYSKGQTS69], name:[zato.security.tls.ca-cert.get-list], SIO request:[{u'cluster_id': 1}]
2015-02-25 22:13:32,172 - INFO - 8016:Dummy-1190 - zato_admin:22 - cid:[K062JEXADBJKNAA58DYYSKGQTS69], name:[zato.security.tls.ca-cert.get-list], response:[{"zato_security_tls_ca_cert_get_list_response": []}]
2015-02-25 22:13:32,174 - INFO - 8016:Dummy-1190 - zato_admin:22 - cid:[K062JEXADBJKNAA58DYYSKGQTS69], name:[zato.service.invoke], response:[{"zato_service_invoke_response": {"response": "eyJ6YXRvX3NlY3VyaXR5X3Rsc19jYV9jZXJ0X2dldF9saXN0X3Jlc3BvbnNlIjogW119\n"}}]
2015-02-25 22:13:32,221 - INFO - 8016:Dummy-1191 - zato_admin:22 - cid:[K04F5YVAQW1D31B2XCFFJJPAR726], name:[zato.service.invoke], SIO request:[{u'name': u'zato.http-soap.get-list', u'data_format': u'json', u'id': u'', u'channel': u'invoke', u'async': False, u'payload': u'eyJjb25uZWN0aW9uIjogImNoYW5uZWwiLCAiY2x1c3Rlcl9pZCI6ICIxIiwgInRyYW5zcG9ydCI6\nICJzb2FwIn0=\n', u'transport': None, u'expiration': 15}]
2015-02-25 22:13:32,223 - INFO - 8016:Dummy-1191 - zato_admin:22 - cid:[K04F5YVAQW1D31B2XCFFJJPAR726], name:[zato.http-soap.get-list], SIO request:[{u'connection': u'channel', u'cluster_id': 1, u'transport': u'soap'}]

$ grep "^2015-02-25 22:13:32" server2/logs/server.log
2015-02-25 22:13:32,250 - ERROR - 8016:Dummy-1191 - zato.http-soap.get-list:22 - Can't run after_handle, e:[Traceback (most recent call last):
2015-02-25 22:13:32,253 - WARNING - 8016:Dummy-1191 - zato.server.service:22 - Could not invoke `zato.http-soap.get-list`, e:`Traceback (most recent call last):
2015-02-25 22:13:32,254 - WARNING - 8016:Dummy-1191 - zato.server.service:22 - Traceback (most recent call last):
2015-02-25 22:13:32,255 - ERROR - 8016:Dummy-1191 - zato.server.connection.http_soap.channel:22 - Caught an exception, cid:[K04F5YVAQW1D31B2XCFFJJPAR726], status_code:[500], _format_exc:[Traceback (most recent call last):
~~~

**Comment by dsuch** [comment]
> since clients cannot be trusted

Naturally. The thing here is that the situation with Chrome uncovered an inconsistency in backend that knows how to reject None names but not '' ones.

Obviously, frontend is not the place to employ such validations.

Backend should be smarter as regarding such input instead of understanding how to reject None only.

**Comment by dsuch** [comment]
Made frontend validation work in Chrome again in commit e64d8ecc1ff0f10f634759513f9d5d8d8179ddd6 - will be released in 2.0.3.

Thanks for reporting it @candlerb!
